Add support for per-ruleset skin layouts

peppy opened the pull request:
This builds off @frenzibyte's original proposal (see #20694) and mostly implements things in a similar way.

Of note, each skin components container can now have both global and per-ruleset components defined. This is done by serialising a new `SkinLayoutInfo` type rather than a naked `IEnumerable<SerialisableDrawableInfo>`. Includes "migration" from previous skins via advanced `try-catch` checking.

This completes a good portion of https://github.com/ppy/osu/issues/22388, but does not yet add updated defaults to, for instance, correct non-matching layouts in legacy skins (mania / taiko).

- [x] Depends on https://github.com/ppy/osu/pull/22648

peppy committed:
Add optional ruleset identifier to `SkinComponentsContainerLookup`

peppy committed:
Always return a non-null container for `SkinComponentsContainerLookup`s

peppy committed:
Change serialisation format of skin layouts to allow more flexibility

Also adds per-ruleset storage for each container type.

peppy committed:
Add ability to retrieve serialisable drawables for specific rulesets

peppy committed:
Use `RulesetInfo` instead of `Ruleset` in skin components lookup

peppy committed:
Update `SkinDeserialisationTest` to work with new serialisation structure

peppy committed:
Ensure skin default component layouts only apply to global layout for now

peppy committed:
Fix `SkinnableLighting` showing up as a user placeable component

peppy committed:
Fix tests which rely on `HUDOverlay`'s `DrawableRuleset` being nullable

bdach committed:
Merge branch 'master' into skin-per-ruleset-layouts

bdach commented on osu.Game/Skinning/SkinComponentsContainerLookup.cs:
```diff
@@ -13,9 +15,14 @@ public class SkinComponentsContainerLookup : ISkinComponentLookup
         public readonly TargetArea Target;
 
-        public SkinComponentsContainerLookup(TargetArea target)
+        public readonly RulesetInfo? Ruleset;
```
```suggestion
        /// <summary>
        /// The ruleset for which skin components should be returned.
        /// A <see langword="null"/> value means that returned components are global and should be applied for all rulesets.
        /// </summary>
        public readonly RulesetInfo? Ruleset;
```

bdach commented on osu.Game/Skinning/SkinComponentsContainerLookup.cs:
```diff
@@ -13,9 +15,14 @@ public class SkinComponentsContainerLookup : ISkinComponentLookup
-        public SkinComponentsContainerLookup(TargetArea target)
+        public readonly RulesetInfo? Ruleset;
+
+        public string GetSerialisableIdentifier() => Ruleset?.ShortName ?? "global";
```
This is a bit weird as it's unused. But if it will be used, I'd prefer that it reuse `SkinLayoutInfo.global_identifier`.

peppy replied:
This was from a previous implementation, will remove for now.

bdach commented on osu.Game/Skinning/SkinLayoutInfo.cs:
```diff
@@ -0,0 +1,36 @@
+    [Serializable]
+    public class SkinLayoutInfo
+    {
+        private const string global_identifier = "global";
```
```suggestion
        private const string global_identifier = @"global";
```

bdach commented on osu.Game/Skinning/Skin.cs:
```diff
@@ -113,18 +114,37 @@ protected Skin(SkinInfo skin, IStorageResourceProvider? resources, IResourceStor
+                    // if that fails, attempt to deserialise using the old naked list.
+                    if (layoutInfo == null)
+                    {
+                        // handle namespace changes...
+                        // can be removed 2023-01-31
+                        jsonContent = jsonContent.Replace(@"osu.Game.Screens.Play.SongProgress", @"osu.Game.Screens.Play.HUD.DefaultSongProgress");
+                        jsonContent = jsonContent.Replace(@"osu.Game.Screens.Play.HUD.LegacyComboCounter", @"osu.Game.Skinning.LegacyComboCounter");
```
Can we remove these while we're here? The deprecation date is up on these anyways.

bdach commented on osu.Game/Skinning/Skin.cs:
```diff
@@ -113,18 +114,37 @@ protected Skin(SkinInfo skin, IStorageResourceProvider? resources, IResourceStor
+                    {
+                    }
+
+                    // if that fails, attempt to deserialise using the old naked list.
```
Are we putting a deprecation date on this pathway, or is it going to have to live forever? Right now it's on the read side, so to get rid of it we'd have to automigrate users' skins somehow I imagine.

peppy replied:
I guess we'll do this one day. I'll note this in the inline comments.

bdach commented on osu.Game/Skinning/SkinComponentsContainer.cs:
```diff
@@ -66,10 +66,10 @@ public void Reload(Container? componentsContainer)
-                return;
-
-            content = componentsContainer;
+            content = componentsContainer ?? new Container
+            {
+                RelativeSizeAxes = Axes.Both
+            };
```
I think with this change, the `content != null` check a bit lower down can be removed? Something like:

```diff
diff --git a/osu.Game/Skinning/SkinComponentsContainer.cs b/osu.Game/Skinning/SkinComponentsContainer.cs
index d4e0b2e69b..d18e9023cd 100644
--- a/osu.Game/Skinning/SkinComponentsContainer.cs
+++ b/osu.Game/Skinning/SkinComponentsContainer.cs
@@ -74,17 +74,12 @@ public void Reload(Container? componentsContainer)
             cancellationSource?.Cancel();
             cancellationSource = null;
 
-            if (content != null)
+            LoadComponentAsync(content, wrapper =>
             {
-                LoadComponentAsync(content, wrapper =>
-                {
-                    AddInternal(wrapper);
-                    components.AddRange(wrapper.Children.OfType<ISerialisableDrawable>());
-                    ComponentsLoaded = true;
-                }, (cancellationSource = new CancellationTokenSource()).Token);
-            }
-            else
+                AddInternal(wrapper);
+                components.AddRange(wrapper.Children.OfType<ISerialisableDrawable>());
                 ComponentsLoaded = true;
+            }, (cancellationSource = new CancellationTokenSource()).Token);
         }
 
         /// <inheritdoc cref="ISerialisableDrawableContainer"/>
```

peppy replied:
Looks good, have applied.

bdach reviewed (commented):
Initial read-through. Haven't done much testing yet.

peppy committed:
Remove unused `GetSerialisableIdentifier` for now

peppy committed:
Add xmldoc for `SkinComponentsContainerLookup.Ruleset`

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy committed:
Mark constant identifier as non-localisable

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy committed:
Add note about skin migrations being on read and remove an older deprecation notice

peppy committed:
Remove unnecesasry null check on `content`

bdach committed:
Clarify comment to avoid playing pronoun game

bdach committed:
Merge branch 'master' into skin-per-ruleset-layouts

bdach committed:
Do not serialise `SkinLayoutInfo.AllDrawables`

- It is entirely derived from `SkinLayoutInfo.DrawableInfo`, which is
  the actual primary thing we want to serialise.
- It will never get read out from any serialised files anyway (corollary
  of the previous point - it is a get-only property derived from another).
- It is only used in tests.

All of the three reasons above make serialising the property out to skin
files nothing more than a waste of space.

bdach commented:
This seems pretty alright, and it doesn't seem to break under some light testing. I did apply a few fixups: one is just a slight comment reword (https://github.com/ppy/osu/pull/22673/commits/43724472c4c7a793c58e57c3083895d5e3e7ee76), while the second is slightly more major but I'm pretty sure it shouldn't be controversial by any means - upon inspecting exported skin files, I noticed that `SkinLayoutInfo.AllDrawables` was being unnecessarily serialised out, so I `[JsonIgnore]`d it.

To speed things up some I'll queue this up for merge now. Let's see where this goes next :+1:
